higgstautau/lephad/filters.py

```diff
@@ -0,0 +1,439 @@
+        #muon TRT requirements AKA the Demon Clause
+        
+        #2 clauses
+        #either in nice eta region with enough hits that are less than 90% outliers
+        #...or...
+        #if outside region, veto events with good hits keep those with crappy hits.
+        event.muons.select(lambda muon:
+            (0.1<abs(muon.eta)<1.9 and
+             (muon.nTRTHits+muon.nTRTOutliers)>5 and
+             muon.nTRTOutliers<0.9*(muon.nTRTHits+muon.nTRTOutliers)) or
+            (( abs(muon.eta)<=0.1 or abs(muon.eta)>=1.9) and
+            not ((muon.nTRTHits+muon.nTRTOutliers)>5 and muon.nTRTOutliers>=0.9*(muon.nTRTHits+muon.nTRTOutliers))))
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class MuonZ0pv(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(MuonZ0pv, self).__init__(**kwargs)
+    def passes(self, event):
+        event.muons.select(lambda muon: muon.trackz0pv<10)
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class MuonCone(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(MuonCone, self).__init__(**kwargs)
+    def passes(self, event):
+        event.muons.select(lambda muon: muon.ptcone40/muon.pt<0.09)
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class ElectronAuthor(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(ElectronAuthor, self).__init__(**kwargs)
+    def passes(self, event):
+        #extra misc electron cuts
+        event.electrons.select(lambda electron: electron.author in (1,3) )
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class ElectronPP(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(ElectronPP, self).__init__(**kwargs)
+    def passes(self, event):
+        event.electrons.select(lambda electron: electron.mediumPP)
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class ElectronOQ(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(ElectronOQ, self).__init__(**kwargs)
+    def passes(self, event):
+        #extra misc electron cuts
+        event.electrons.select(lambda electron: (electron.OQ&1446)==0)
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class ElectronCone(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(ElectronCone, self).__init__(**kwargs)
+    def passes(self, event):
+        #percentage energy in cone. electron.cl_E/cosh(electron.tracketa) is electron pt
+        event.electrons.select(lambda electron: electron.ptcone40/(electron.cl_E/cosh(electron.tracketa))<0.17)
+        return len(event.electrons)+len(event.muons) >= self.min_leps
+
+class Triggers(EventFilter):
```

Comment:
yes. After I discuss with Clara and Mohamad we can each adjust the trigger specifics for the respective modules.